cpp/src/copying/contiguous_split.cu

```diff
@@ -762,6 +789,142 @@ struct size_of_helper {
+    0, [num_bufs, num_chunks = num_chunks_func{chunks.begin()}] __device__(size_type i) {
+      return i == num_bufs ? 0 : num_chunks(i);
+    });
+  thrust::exclusive_scan(rmm::exec_policy(stream),
+                         buf_count_iter,
+                         buf_count_iter + num_bufs + 1,
+                         chunk_offsets.begin(),
+                         0);
+
+  auto out_to_in_index = [chunk_offsets = chunk_offsets.begin(), num_bufs] __device__(size_type i) {
+    return static_cast<size_type>(
+             thrust::upper_bound(thrust::seq, chunk_offsets, chunk_offsets + num_bufs + 1, i) -
+             chunk_offsets) -
+           1;
+  };
+
+  // apply the chunking.
+  auto const num_chunks =
+    cudf::detail::make_counting_transform_iterator(0, num_chunks_func{chunks.begin()});
+  size_type const new_buf_count =
+    thrust::reduce(rmm::exec_policy(stream), num_chunks, num_chunks + chunks.size());
+  rmm::device_uvector<dst_buf_info> d_dst_buf_info(new_buf_count, stream);
+  auto iter = thrust::make_counting_iterator(0);
+  thrust::for_each(rmm::exec_policy(stream),
+                   iter,
+                   iter + new_buf_count,
+                   [_d_dst_buf_info,
+                    d_dst_buf_info = d_dst_buf_info.begin(),
+                    chunks         = chunks.begin(),
+                    chunk_offsets  = chunk_offsets.begin(),
+                    num_bufs,
+                    num_src_bufs,
+                    out_to_in_index] __device__(size_type i) {
+                     size_type const in_buf_index = out_to_in_index(i);
+                     size_type const chunk_index  = i - chunk_offsets[in_buf_index];
+                     auto const chunk_size        = thrust::get<1>(chunks[in_buf_index]);
+                     dst_buf_info const& in       = _d_dst_buf_info[in_buf_index];
+
+                     // adjust info
+                     dst_buf_info& out = d_dst_buf_info[i];
+                     out.element_size  = in.element_size;
+                     out.value_shift   = in.value_shift;
+                     out.bit_shift     = in.bit_shift;
+                     out.valid_count =
+                       in.valid_count;  // valid count will be set to 1 if this is a validity buffer
+                     out.src_buf_index = in.src_buf_index;
+                     out.dst_buf_index = in.dst_buf_index;
+
+                     size_type const elements_per_chunk =
+                       out.element_size == 0 ? 0 : chunk_size / out.element_size;
+                     out.num_elements = ((chunk_index + 1) * elements_per_chunk) > in.num_elements
+                                          ? in.num_elements - (chunk_index * elements_per_chunk)
+                                          : elements_per_chunk;
+
+                     size_type const rows_per_chunk =
+                       out.valid_count > 0 ? elements_per_chunk * 32 : elements_per_chunk;
```

Comment:
32 here seems magical. Why is it 32? constexpr?

Reply:
It's the # of rows per element in a validity buffer (where element is bitmask_type). I'll add some comments.

Reply:
How about using [`size_in_bits<bitmask_type>()` ](https://github.com/rapidsai/cudf/blob/07fa8883ae3ad8a8614eba8eca8b2fda79bc5e30/cpp/include/cudf/utilities/bit.hpp#L45)